tunnel/intra/sni_reporter.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intra
+
+import (
+	"io"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/choir"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/tunnel/intra/doh"
+	"github.com/eycorsican/go-tun2socks/common/log"
+)
+
+// Number of bins to assign reports to.  Should be large enough for
+// k-anonymity goals.  See the Choir documentation for more info.
+const bins = 32
+
+// Number of values in each report.  The two values are
+// * success/failure
+// * timeout/close
+const values = 2
+
+// Burst duration.  Only one report will be sent in each interval
+// to avoid correlated reports.
+const burst = 10 * time.Second
+
+// tcpSNIReporter is a thread-safe wrapper around choir.Reporter
+type tcpSNIReporter struct {
+	mu     sync.RWMutex // Protects dns, suffix, and r.
+	dns    doh.Transport
+	suffix string
+	r      choir.Reporter
+}
+
+// SetDNS changes the DNS transport used for uploading reports.
+func (r *tcpSNIReporter) SetDNS(dns doh.Transport) {
+	r.mu.Lock()
+	r.dns = dns
+	r.mu.Unlock()
+}
+
+// Send implements choir.ReportSender.
+func (r *tcpSNIReporter) Send(report choir.Report) error {
+	r.mu.RLock()
+	suffix := r.suffix
+	dns := r.dns
+	r.mu.RUnlock()
+	q, err := choir.FormatQuery(report, suffix)
+	if err != nil {
+		log.Warnf("Failed to construct query for Choir: %v", err)
+		return nil
```

Comment:
Should you return `err`? Same below.

Reply:
The only action Choir would take on an error is to print it using the standard Go logging framework: https://github.com/Jigsaw-Code/choir/blob/master/client.go#L105.  Since we have our own logging framework here, I figured it would be better to log here and not there.

Reply:
Makes sense 👍